### PR TITLE
matching: introduce consistent hashing matcher

### LINKED ISSUE #14782: Add consistent hashing InputMatcher
Tracks adding a simple InputMatcher that takes a configurable threshold and will consistently hash the input string, matching if the resulting hash value is above the configured threshold:

```
match = (hash(input) % 100) > threshold
```

This would allow users to configure a match rule that will consistently match X% of requests based on a random input string (e.g. a tracing header).

### PR DESCRIPTION
**Pull request by snowp**
This introduces a new matcher that allows matching on an input value by
computing a hash value and matching if the value % (configured value) is
greater than a configured  threshold. This is useful in being able to
define match criteria that should match for a certain % of input values
in a way that is consistent between independent Envoy instances (e.g. it
does not rely on a random input).

Risk Level: Low, new extension
Testing: UTs
Docs Changes: Inline proto docs
Release Notes: n/a
Platform Specific Features: n/a
Fixes #14782

### CODE REVIEW

**Comment by snowp** [comment]
@donyu

**Comment by snowp** [comment]
Will split this into a PR for the changes to the core matching logic (since I want to make similar changes to the other factories) and leave this as just the extension change, hence the draft

File: api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// [#protodoc-title: Consistent Hashing Matcher]
+// [#extension: envoy.matching.input_matchers.consistent_hashing]
+
+// The consistent hashing matchers computes a consistent hash from the input and matches if the resulting hash
+// is within the configured threshold.
+// More specifically, this matcher evaluates to true if hash(input) % modulo > threshold.
+message ConsistentHashing {
+  // The threshold the resulting hash must be over in order for this matcher to evaluate to true.
+  // This value must be below the configured modulo value.
+  uint32 threshold = 1 [(validate.rules).uint32 = {gt: 0}];
```
Comment (htuch) [inline api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto:24]: Could this be zero to mean 100%?

File: source/extensions/matching/input_matchers/consistent_hashing/matcher.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+public:
+  Matcher(uint32_t threshold, uint32_t modulo) : threshold_(threshold), modulo_(modulo) {}
+  bool match(absl::optional<absl::string_view> input) override {
+    // Only match if the value is present.
+    if (!input) {
+      return false;
+    }
+
+    // Otherwise, match if (hash(input) % modulo) > threshold.
+    return HashUtil::xxHash64(*input) % modulo_ > threshold_;
```
Comment (htuch) [inline source/extensions/matching/input_matchers/consistent_hashing/matcher.h:23]: One interesting implication of the consistent hasher API guarantee that the results are fleet-wide consistent is that if we ever change `xxHash64`, or `xxhash` itself internally makes some inconsistent breaking change, the property will be violated during rollouts. FWIW we have this problem today with our affinity load balancers. Not sure how you prefer to handle it (maybe warn in API documentation is enough?) but worth considering.

File: test/extensions/matching/input_matchers/consistent_hashing/matcher_test.cc [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+namespace Envoy {
+namespace Extensions {
+namespace Matching {
+namespace InputMatchers {
+namespace ConsistentHashing {
+
+TEST(MatcherTest, EmptyValue) {
+  Matcher matcher(10, 100);
+
+  ASSERT_FALSE(matcher.match(absl::nullopt));
```
Comment (htuch) [inline test/extensions/matching/input_matchers/consistent_hashing/matcher_test.cc:14]: Nit: these can all be `EXPECT`.

File: test/extensions/matching/input_matchers/consistent_hashing/matcher_test.cc [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  Matcher matcher(59, 100);
+
+  // Changing the threshold to 59 means that we no longer match.
+  ASSERT_FALSE(matcher.match("hello"));
+}
+} // namespace ConsistentHashing
+} // namespace InputMatchers
+} // namespace Matching
+} // namespace Extensions
+} // namespace Envoy
```
Comment (htuch) [inline test/extensions/matching/input_matchers/consistent_hashing/matcher_test.cc:36]: Maybe there are some stronger tests, e.g. building two fresh instances should always agree.

**Review by htuch — COMMENTED** [review]
Looks like a useful addition. How come this requirement never come up in RouteConfiguration matchers?

File: CODEOWNERS [hunk 5]
```diff
@@ -162,3 +162,5 @@ extensions/filters/http/oauth2 @rgs1 @derekargueta @snowp
 /*/extensions/filters/http/kill_request @qqustc @htuch
 # Rate limit expression descriptor
 /*/extensions/rate_limit_descriptors/expr @kyessenov @lizan
+# hash input matcher
+/*/extensions/matching/input_matchers/consistent_hashing @snowp @donyu
```
Comment (donyu) [inline CODEOWNERS:166]: 👍

File: source/extensions/matching/input_matchers/consistent_hashing/matcher.h (hunk 2 above)
Comment (snowp) [reply source/extensions/matching/input_matchers/consistent_hashing/matcher.h:23]: I'll add a warning to the docs, not sure what else we could do here besides implementing our own hash that we promise we'll never change (or never update xxhash?).

File: api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto (hunk 1 above)
Comment (snowp) [reply api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto:24]: Yeah good point, will remove the validation

**Comment by sschepens** [comment]
> Looks like a useful addition. How come this requirement never come up in RouteConfiguration matchers? 

@htuch We're actually in need of this kind of matching in RouteConfiguration and are using Lua to get it done.

**Comment by sschepens** [comment]
@snowp @htuch do you think it's possible to also add a configurable "seed", this is something we also need, because without a seed a given value would consistently match across every given service using a rule.

We're using this kind of hashing for "sticky" deployments based on a unique device identificator header, what happens without a seed is when a value of the header becomes covered by the threshold, then that user would simultaneously be enabled on all sticky deployments on the platform.

**Comment by snowp** [comment]
@sschepens Sure, I was going to add that in later as we'll need this as well. Would simply passing through a uint64_t to xxhash be sufficient for your use case?

**Comment by sschepens** [comment]
@snowp yep, i think that would be alright, any string seed could be converted by management servers to a uint64

**Comment by sschepens** [comment]
@snowp another question, will this be usable from RouteConfiguration or is this part of something else?

File: source/extensions/matching/input_matchers/consistent_hashing/matcher.h (hunk 2 above)
Comment (htuch) [reply source/extensions/matching/input_matchers/consistent_hashing/matcher.h:23]: Can we model 100% match here given the `>`? Should it be `>=`?

File: test/extensions/matching/input_matchers/consistent_hashing/config_test.cc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+)EOF";
+
+  envoy::config::core::v3::TypedExtensionConfig config;
+  TestUtility::loadFromYaml(yaml_string, config);
+
+  ConsistentHashingConfig factory;
+  auto message = Config::Utility::translateAnyToFactoryConfig(
+      config.typed_config(), ProtobufMessage::getStrictValidationVisitor(), factory);
+  auto matcher = factory.createInputMatcher(*message, context);
+  ASSERT_NE(nullptr, matcher);
```
Comment (htuch) [inline test/extensions/matching/input_matchers/consistent_hashing/config_test.cc:31]: Nit: `EXPECT` for a final assertion (even if fatal).

**Review by htuch — COMMENTED** [review]
Looks great. Just one question.

**Comment by snowp** [comment]
@htuch ptal

File: source/extensions/matching/input_matchers/consistent_hashing/matcher.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+public:
+  Matcher(uint32_t threshold, uint32_t modulo, uint64_t seed)
+      : threshold_(threshold), modulo_(modulo), seed_(seed) {}
+  bool match(absl::optional<absl::string_view> input) override {
+    // Only match if the value is present.
+    if (!input) {
+      return false;
+    }
+
+    // Otherwise, match if (hash(input) % modulo) > threshold.
```
Comment (htuch) [inline source/extensions/matching/input_matchers/consistent_hashing/matcher.h:23]: Nit: `>=`

File: api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+option java_outer_classname = "ConsistentHashingProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Consistent Hashing Matcher]
+// [#extension: envoy.matching.input_matchers.consistent_hashing]
+
+// The consistent hashing matchers computes a consistent hash from the input and matches if the resulting hash
+// is within the configured threshold.
+// More specifically, this matcher evaluates to true if hash(input) % modulo > threshold.
```
Comment (htuch) [inline api/envoy/extensions/matching/input_matchers/consistent_hashing/v3/consistent_hashing.proto:20]: Nit: `>=`

**Review by htuch — COMMENTED** [review]
LGTM modulo nits.

**Comment by snowp** [comment]
@htuch ptal

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**